### PR TITLE
Implement adaptive spin lock on ARM Apple devices

### PR DESCRIPTION
**Pull request by mrsaturnsan**
Updated spin lock to use adaptive synchronization as suggested. Normally unfair lock will block, which defeats the purpose of spin locks.

"As it turns out os_unfair_lock doesn’t spin on contention, it makes the calling thread [sleep right away when it finds a contended lock](https://github.com/apple/darwin-libplatform/blob/215b09856ab5765b7462a91be7076183076600df/src/os/lock.c#L536)."

https://hacks.mozilla.org/2022/10/improving-firefox-responsiveness-on-macos/

### CODE REVIEW

**Comment by clayjohn** [comment]
CC @stuartcarnie Who added `os_unfair_lock`

**Comment by stuartcarnie** [comment]
Thanks for the PR, @mrsaturnsan!

Unfortunately, this is a private API, which could affect our iOS deployments and possibly result in rejected builds to the App Store.

Further, the original post was on Hacker News, and a fellow software engineer whom I trust, [Saagar Jha](https://saagarjha.com) posted [this comment](https://news.ycombinator.com/item?id=33157142) about the API, and warns against using it, as it is private for a reason. I have previously spoken to Apple engineers about the `os_unfair_lock` API, and they say it is generally the right approach to replace a spin lock on Apple platforms.

**Comment by stuartcarnie** [comment]
@mrsaturnsan excellent – if you have used it, and Apple has made the API public in the latest OS, that is certainly an endorsement.

**Comment by stuartcarnie** [comment]
@clayjohn LGTM!

**Comment by clayjohn** [comment]
AFAIK we support as far back as MacOS 10.13 and iOS 12.0, would using those older version be a problem with this PR?

CC @bruvzg

**Comment by bruvzg** [comment]
> AFAIK we support as far back as MacOS 10.13 and iOS 12.0, would using those older version be a problem with this PR?

We do, supporting both probably can be done using `if (__builtin_available(macOS 15, iOS 18, *)) { ... }` (since it's only public since macOS 15 / iOS 18, it probably should be limited to these versions).

**Comment by bruvzg** [comment]
> The problem with if (__builtin_available(macOS 15, iOS 18, *)) { ... } is that it's a runtime check so using the new function will cause compile errors when targeting older OS versions. 

Usually compiling with old compiler is not important, we only want to support running on old OSes. And we already have `@available` check for macOS 13.4 and iOS 16. But I guess if it's tested on 10.15, `available` check can be 10.15 as well (not sure which iOS versions are supported).

**Comment by bruvzg** [comment]
> I think this would be the safest option.

Yes, it should be OK.

File: core/os/spin_lock.h [hunk 1]
```diff
@@ -37,12 +37,26 @@
...
+OS_UNFAIR_LOCK_AVAILABILITY OS_EXPORT OS_NOTHROW OS_NONNULL_ALL void os_unfair_lock_lock_with_options(os_unfair_lock_t lock, os_unfair_lock_options_t options);
+}
+#endif
+
 class SpinLock {
 	mutable os_unfair_lock _lock = OS_UNFAIR_LOCK_INIT;
 
 public:
 	_ALWAYS_INLINE_ void lock() const {
+#if __arm64__ && TARGET_OS_OSX
```
Comment (bruvzg) [inline core/os/spin_lock.h:55]:
```suggestion
#if defined(__aarch64__) && defined(MACOS_ENABLED)
```
Just to use same defines the rest of the code do.

**Review by bruvzg — APPROVED** [review]
Please squash the commits, see [Pull request workflow](https://docs.godotengine.org/en/stable/contributing/workflow/pr_workflow.html#the-interactive-rebase).

File: core/os/spin_lock.h [hunk 2]
```diff
@@ -37,12 +37,26 @@
 
 #include <os/lock.h>
 
+#if defined(__aarch64__) && defined(MACOS_ENABLED)
+#define OS_UNFAIR_LOCK_DATA_SYNCHRONIZATION (0x00010000)
+#define OS_UNFAIR_LOCK_ADAPTIVE_SPIN (0x00040000)
```
Comment (RandomShaper) [inline core/os/spin_lock.h:42]: We're preferring first-class constants where possible;
```suggestion
static constexpr uint32_t OS_UNFAIR_LOCK_DATA_SYNCHRONIZATION = 0x00010000;
static constexpr uint32_t OS_UNFAIR_LOCK_ADAPTIVE_SPIN = 0x00040000;
```

Would there be any issue in this case?

**Comment by clayjohn** [comment]
Will need to be squashed again

**Comment by RandomShaper** [comment]
Some time ago, I added a benchmark that works as an indirect test of `SpinLock`: https://github.com/godotengine/godot-benchmarks/pull/95. I've used it in my own non-Mac spin-lock improvement PR: https://github.com/godotengine/godot/pull/85167.

It'd be amazing if we could get some numbers from a test based on that (maybe with different amounts of threads) with the following `SpinLock` implementations, on some ARM-based Mac or iPhone:
- Current without Apple-specific implementation.
- The same, from #85167.
- Apple-specific before this PR.
- Apple-specific after this PR.

I know it's a lot to ask, so if we could at least compare before'n'after this PR on MacOS, that'd be enough here.

**Comment by clayjohn** [comment]
Maybe the TPS demo? https://github.com/godotengine/tps-demo I suspect that it won't have much thread contention as it is mostly single threaded though

**Comment by stuartcarnie** [comment]
> Normally unfair lock will block, which defeats the purpose of spin locks.

They don't block if they are not contested, which is probably likely under normal conditions. This is also evident in the single threaded benchmark result, where both are almost identical in performance. 

I reformatted your results into a table, so they are easier to read and compare (I don't know the time units):

| Benchmark                                   | Before | After  | ± change |
|:--------------------------------------------|:-------|:-------|----------|
| core/object_db/single                       | 18.65  | 17.71  | 1.05     |
| core/object_db/2_threads_full_contention    | 59.71  | 51.29  | 1.16     |
| core/object_db/2_threads_half_contention    | 38.24  | 34.03  | 1.12     |
| core/object_db/2_threads_little_contention  | 22.54  | 22.25  | 1.01     |
| core/object_db/4_threads_full_contention    | 180.6  | 173.6  | 1.04     |
| core/object_db/4_threads_half_contention    | 101.4  | 95.34  | 1.06     |
| core/object_db/4_threads_little_contention  | 35.11  | 32.9   | 1.07     |
| core/object_db/4_threads_slope_contention   | 135.5  | 127.7  | 1.06     |
| core/object_db/8_threads_full_contention    | 516.4  | 785.5  | 0.66     |
| core/object_db/8_threads_half_contention    | 299.4  | 461.2  | 0.65     |
| core/object_db/8_threads_little_contention  | 74.33  | 102.2  | 0.73     |
| core/object_db/8_threads_slope_contention   | 314.4  | 410.2  | 0.77     |
| core/object_db/12_threads_full_contention   | 1317.0 | 1991.0 | 0.66     |
| core/object_db/12_threads_half_contention   | 643.7  | 1055.0 | 0.61     |
| core/object_db/12_threads_little_contention | 138.5  | 204.6  | 0.68     |
| core/object_db/12_threads_slope_contention  | 586.1  | 1017.0 | 0.58     |

My interpretation is that it is slightly better when there are few threads (2 - 4 threads), but as the thread count increases, the worst case is much worse than the current implementation. This is likely a combination of number of available CPUs, and overall system activity, and so blocking is better, as it forces the thread to yield sooner to give the other threads a chance to make progress.

I suspect if you have a 16 core Mac, you might see the improved performance extend to 8 and possibly even 12 threads.

## TL;DR

I'd argue that although best case is slightly better, worst case is significantly worse.
